File: tests/modules/login.module.js
/**
 * 登录模块
 * 包含登录、租户名验证、语言切换功能
 */

/**
 * 执行登录操作
 * @param {Page} page - Playwright页面对象
 * @param {Object} credentials - 登录凭据
 * @param {string} credentials.account - 账号
 * @param {string} credentials.password - 密码
 * @param {string} credentials.brandAlias - 品牌别名
 */
async function login(page, { account = 'admin', password = 'admin@123', brandAlias = 'hex' }) {
  // 步骤1: 打开登录页面
  await page.goto('https://saas-auth-qa.hexcloud.cn/page/login');
  
  // 使用更宽松的等待策略
  try {
    await page.waitForLoadState('domcontentloaded', { timeout: 10000 });
  } catch (e) {
    console.log('等待domcontentloaded超时，继续执行:', e.message);
  }
  await page.waitForTimeout(2000);

  // 步骤2: 填写登录表单
  await page.waitForSelector('input[type="text"], input[type="password"]', { timeout: 10000 });
  
  // 填写账号
  const accountSelectors = [
    'input[name="account"]',
    'input[name="username"]',
    'input[id="account"]',
    'input[id="username"]',
    'input[placeholder*="账号"]',
    'input[placeholder*="用户名"]',
    'input[type="text"]'
  ];
  
  let accountInput = null;
  for (const selector of accountSelectors) {
    try {
      accountInput = await page.locator(selector).first();
      if (await accountInput.isVisible({ timeout: 2000 })) {
        break;
      }
    } catch (e) {
      continue;
    }
  }
  
  if (accountInput) {
    await accountInput.fill(account);
  } else {
    throw new Error('无法找到账号输入框');
  }

  // 填写密码
  const passwordSelectors = [
    'input[name="password"]',
    'input[id="password"]',
    'input[type="password"]'
  ];
  
  let passwordInput = null;
  for (const selector of passwordSelectors) {
    try {
      passwordInput = await page.locator(selector).first();
      if (await passwordInput.isVisible({ timeout: 2000 })) {
        break;
      }
    } catch (e) {
      continue;
    }
  }
  
  if (passwordInput) {
    await passwordInput.fill(password);
  } else {
    throw new Error('无法找到密码输入框');
  }

  // 填写品牌别名
  const brandSelectors = [
    'input[name*="brand"]',
    'input[name*="alias"]',
    'input[id*="brand"]',
    'input[id*="alias"]',
    'input[placeholder*="品牌"]',
    'input[placeholder*="别名"]'
  ];
  
  let brandInput = null;
  for (const selector of brandSelectors) {
    try {
      brandInput = await page.locator(selector).first();
      if (await brandInput.isVisible({ timeout: 2000 })) {
        break;
      }
    } catch (e) {
      continue;
    }
  }
  
  if (brandInput) {
    await brandInput.fill(brandAlias);
  } else {
    console.warn('未找到品牌别名输入框，可能不需要填写');
  }

  // 步骤3: 勾选用户协议并提交登录表单
  const agreementSelectors = [
    'input[type="checkbox"]',
    'checkbox',
    '[role="checkbox"]'
  ];
  
  let agreementCheckbox = null;
  for (const selector of agreementSelectors) {
    try {
      agreementCheckbox = page.locator(selector).first();
      if (await agreementCheckbox.isVisible({ timeout: 2000 })) {
        const isChecked = await agreementCheckbox.isChecked();
        if (!isChecked) {
          await agreementCheckbox.check();
        }
        break;
      }
    } catch (e) {
      continue;
    }
  }
  
  await page.waitForTimeout(500);
  
  // 点击登录按钮
  const buttonSelectors = [
    'button:has-text("登录")',
    'button[type="submit"]',
    'button:has-text("Login")',
    'input[type="submit"]',
    'button.btn-primary',
    'button.login-btn'
  ];
  
  let loginButton = null;
  for (const selector of buttonSelectors) {
    try {
      loginButton = await page.locator(selector).first();
      if (await loginButton.isVisible({ timeout: 2000 })) {
        break;
      }
    } catch (e) {
      continue;
    }
  }
  
  if (loginButton) {
    console.log('找到登录按钮，准备点击');
    // 等待按钮可点击
    await loginButton.waitFor({ state: 'visible', timeout: 5000 });
    await page.waitForTimeout(500);
    
    // 点击登录按钮并等待导航
    await Promise.all([
      page.waitForURL((url) => !url.includes('/page/login'), { timeout: 20000 }).catch(() => {}),
      loginButton.click()
    ]);
  } else {
    console.log('未找到登录按钮，尝试按Enter键');
    await page.keyboard.press('Enter');
  }
  
  // 等待页面跳转，检查是否登录成功
  let loginSuccess = false;
  const maxWaitTime = 20000; // 最多等待20秒
  const startTime = Date.now();
  
  while (Date.now() - startTime < maxWaitTime) {
    await page.waitForTimeout(1000);
    const currentUrl = page.url();
    
    if (!currentUrl.includes('/page/login')) {
      loginSuccess = true;
      console.log('登录成功，页面已跳转到:', currentUrl);
      break;
    }
    
    // 检查是否有错误提示
    try {
      const pageText = await page.textContent('body').catch(() => '');
      if (pageText && (pageText.includes('错误') || pageText.includes('失败') || pageText.includes('error') || pageText.includes('Error'))) {
        console.log('检测到错误信息:', pageText.substring(0, 200));
        throw new Error('登录失败，页面显示错误信息');
      }
    } catch (e) {
      if (e.message.includes('登录失败')) {
        throw e;
      }
    }
  }
  
  if (!loginSuccess) {
    const finalUrl = page.url();
    const pageText = await page.textContent('body').catch(() => '');
    console.log('登录超时，当前URL:', finalUrl);
    console.log('页面内容预览:', pageText ? pageText.substring(0, 500) : '空内容');
    throw new Error('登录失败，页面仍在登录页面或登录超时');
  }
  
  // 使用更宽松的等待策略
  try {
    await page.waitForLoadState('domcontentloaded', { timeout: 10000 });
  } catch (e) {
    console.log('等待domcontentloaded超时，继续执行:', e.message);
  }
  
  await page.waitForTimeout(3000);
  console.log('登录后页面URL:', page.url());
}

/**
 * 验证租户名
 * @param {Page} page - Playwright页面对象
 * @param {string} expectedTenantName - 期望的租户名，默认为"合阔x"
 */
async function verifyTenantName(page, expectedTenantName = '合阔x') {
  await page.waitForLoadState('domcontentloaded');
  await page.waitForTimeout(2000);
  
  try {
    const pageUrl = page.url();
    console.log('当前页面URL:', pageUrl);
    
    await page.waitForLoadState('domcontentloaded');
    const pageTitle = await page.title().catch(() => '无法获取标题');
    console.log('当前页面标题:', pageTitle);
  } catch (e) {
    console.log('获取页面信息失败:', e.message);
  }
  
  let tenantElement = null;
  let found = false;
  
  try {
    tenantElement = page.getByText(expectedTenantName, { exact: false }).first();
    await tenantElement.waitFor({ state: 'visible', timeout: 5000 });
    found = true;
  } catch (e) {
    // 继续尝试其他方法
  }
  
  if (!found) {
    try {
      const headerSelectors = ['header', 'nav', '[class*="header"]', '[class*="navbar"]'];
      for (const headerSelector of headerSelectors) {
        try {
          const header = page.locator(headerSelector).first();
          if (await header.isVisible({ timeout: 2000 })) {
            tenantElement = header.getByText(expectedTenantName, { exact: false }).first();
            if (await tenantElement.isVisible({ timeout: 2000 })) {
              found = true;
              break;
            }
          }
        } catch (e) {
          continue;
        }
      }
    } catch (e) {
      console.log('在header中查找失败');
    }
  }
  
  if (!found) {
    const pageText = await page.textContent('body');
    if (pageText && pageText.includes(expectedTenantName)) {
      found = true;
      console.log('租户名存在于页面文本中');
    }
  }
  
  if (!found) {
    throw new Error(`未找到租户名: ${expectedTenantName}`);
  }
  
  console.log(`✓ 租户名验证通过: ${expectedTenantName}`);
}

/**
 * 检查并切换语言设置为简体中文
 * @param {Page} page - Playwright页面对象
 */
async function switchToSimplifiedChinese(page) {
  // 使用更宽松的等待策略，避免超时
  try {
    await page.waitForLoadState('domcontentloaded', { timeout: 10000 });
  } catch (e) {
    console.log('等待domcontentloaded超时，继续执行:', e.message);
  }
  await page.waitForTimeout(2000);
  
  const pageText = await page.textContent('body');
  const pageInnerText = await page.evaluate(() => document.body.innerText);
  
  const englishKeywords = [
    'Store operations',
    'Warehouse operations', 
    'BOH Supply Chain Platform',
    'Organizational management',
    'Product management',
    'Formula management'
  ];
  
  const chineseKeywords = [
    '门店运营',
    'BOH供应链中台',
    '组织管理',
    '产品管理',
    '配方管理',
    '仓库运营'
  ];
  
  let hasEnglishMenu = false;
  let hasChineseMenu = false;
  
  for (const keyword of englishKeywords) {
    if (pageText && pageText.includes(keyword)) {
      hasEnglishMenu = true;
      break;
    }
  }
  
  for (const keyword of chineseKeywords) {
    if (pageText && pageText.includes(keyword)) {
      hasChineseMenu = true;
      break;
    }
  }
  
  console.log('检测到英文菜单:', hasEnglishMenu);
  console.log('检测到中文菜单:', hasChineseMenu);
  
  if (hasEnglishMenu && !hasChineseMenu) {
    console.log('需要切换到简体中文');
    
    let languageSwitched = false;
    
    // 方法1: 查找页面右上角的语言切换控件
    try {
      const topRightSelectors = [
        '[class*="header"] [class*="language"]',
        '[class*="header"] [class*="lang"]',
        '[class*="navbar"] [class*="language"]',
        '[class*="user"] [class*="language"]',
        'header [class*="language"]',
        'header [class*="lang"]'
      ];
      
      for (const selector of topRightSelectors) {
        try {
          const langElement = page.locator(selector).first();
          if (await langElement.isVisible({ timeout: 3000 })) {
            await langElement.click();
            await page.waitForTimeout(1000);
            
            const chineseOption = page.locator('text=/简体中文|中文|Chinese|zh-CN|zh_CN/i').first();
            if (await chineseOption.isVisible({ timeout: 2000 })) {
              await chineseOption.click();
              await page.waitForTimeout(2000);
              languageSwitched = true;
              console.log('通过右上角语言控件切换成功');
              break;
            }
          }
        } catch (e) {
          continue;
        }
      }
    } catch (e) {
      console.log('右上角语言控件查找失败');
    }
    
    // 方法2: 查找所有语言相关的选择器
    if (!languageSwitched) {
      const languageSelectors = [
        '[class*="language"]',
        '[class*="lang"]',
        '[id*="language"]',
        '[id*="lang"]',
        'select[name*="language"]',
        'select[name*="lang"]',
        '[aria-label*="语言"]',
        '[aria-label*="Language"]',
        '[title*="语言"]',
        '[title*="Language"]',
        '[data-testid*="language"]',
        '[data-testid*="lang"]'
      ];
      
      for (const selector of languageSelectors) {
        try {
          const langElements = await page.locator(selector).all();
          for (const langElement of langElements) {
            if (await langElement.isVisible({ timeout: 2000 })) {
              const tagName = await langElement.evaluate(el => el.tagName);
              
              if (tagName === 'SELECT') {
                await langElement.selectOption({ label: /简体中文|中文|Chinese|zh-CN|zh_CN/i });
                await page.waitForTimeout(2000);
                languageSwitched = true;
                console.log('通过下拉框切换语言成功');
                break;
              } else {
                await langElement.click();
                await page.waitForTimeout(1000);
                
                const chineseOption = page.locator('text=/简体中文|中文|Chinese|zh-CN|zh_CN/i').first();
                if (await chineseOption.isVisible({ timeout: 2000 })) {
                  await chineseOption.click();
                  await page.waitForTimeout(2000);
                  languageSwitched = true;
                  console.log('通过点击语言控件切换成功');
                  break;
                }
              }
            }
          }
          if (languageSwitched) break;
        } catch (e) {
          continue;
        }
      }
    }
    
    // 等待语言切换生效
    if (languageSwitched) {
      // 使用更宽松的等待策略
      try {
        await page.waitForLoadState('domcontentloaded', { timeout: 10000 });
      } catch (e) {
        console.log('等待domcontentloaded超时，继续执行:', e.message);
      }
      await page.waitForTimeout(3000);
      
      const newPageText = await page.textContent('body');
      let hasChineseNow = false;
      
      for (const keyword of chineseKeywords) {
        if (newPageText && newPageText.includes(keyword)) {
          hasChineseNow = true;
          break;
        }
      }
      
      if (hasChineseNow) {
        console.log('语言切换成功，菜单已显示为简体中文');
      } else {
        console.log('警告: 语言切换可能未生效，继续执行');
      }
    } else {
      console.log('警告: 未能找到语言切换控件，继续执行（菜单可能仍为英文）');
    }
  } else if (hasChineseMenu) {
    console.log('菜单已显示为简体中文，无需切换');
  } else {
    console.log('无法确定菜单语言，继续执行');
  }
}

module.exports = {
  login,
  verifyTenantName,
  switchToSimplifiedChinese
};

File: tests/modules/login_module.py
"""
登录模块
包含登录、租户名验证功能
"""

from playwright.sync_api import Page
from ..config.login_config import LOGIN_URL, CREDENTIALS


def login(
    page: Page,
    login_url: str = LOGIN_URL,
    account: str = None,
    password: str = None,
    brand_alias: str = None
):
    """
    执行登录操作
    
    Args:
        page: Playwright页面对象
        login_url: 登录页面URL（可选，默认使用配置文件中的URL）
        account: 账号（可选，默认使用配置文件中的账号）
        password: 密码（可选，默认使用配置文件中的密码）
        brand_alias: 品牌别名（可选，默认使用配置文件中的品牌别名）
    """
    # 使用默认凭据
    if account is None:
        account = CREDENTIALS['account']
    if password is None:
        password = CREDENTIALS['password']
    if brand_alias is None:
        brand_alias = CREDENTIALS['brandAlias']
    
    # 步骤1: 打开登录页面
    page.goto(login_url)
    
    # 使用更宽松的等待策略
    try:
        page.wait_for_load_state('domcontentloaded', timeout=10000)
    except Exception as e:
        print(f'等待domcontentloaded超时，继续执行: {e}')
    page.wait_for_timeout(2000)

    # 步骤2: 填写登录表单
    page.wait_for_selector('input[type="text"], input[type="password"]', timeout=10000)
    
    # 填写账号
    account_selectors = [
        'input[name="account"]',
        'input[name="username"]',
        'input[id="account"]',
        'input[id="username"]',
        'input[placeholder*="账号"]',
        'input[placeholder*="用户名"]',
        'input[type="text"]'
    ]
    
    account_input = None
    for selector in account_selectors:
        try:
            account_input = page.locator(selector).first
            if account_input.is_visible(timeout=2000):
                break
        except Exception:
            continue
    
    if account_input:
        account_input.fill(account)
    else:
        raise Exception('无法找到账号输入框')

    # 填写密码
    password_selectors = [
        'input[name="password"]',
        'input[id="password"]',
        'input[type="password"]'
    ]
    
    password_input = None
    for selector in password_selectors:
        try:
            password_input = page.locator(selector).first
            if password_input.is_visible(timeout=2000):
                break
        except Exception:
            continue
    
    if password_input:
        password_input.fill(password)
    else:
        raise Exception('无法找到密码输入框')

    # 填写品牌别名
    brand_selectors = [
        'input[name*="brand"]',
        'input[name*="alias"]',
        'input[id*="brand"]',
        'input[id*="alias"]',
        'input[placeholder*="品牌"]',
        'input[placeholder*="别名"]'
    ]
    
    brand_input = None
    for selector in brand_selectors:
        try:
            brand_input = page.locator(selector).first
            if brand_input.is_visible(timeout=2000):
                break
        except Exception:
            continue
    
    if brand_input:
        brand_input.fill(brand_alias)
        print(f'已填写品牌别名: {brand_alias}')
    else:
        print('未找到品牌别名输入框，继续执行')

    # 检查并勾选协议复选框（如果需要）
    try:
        agreement_selectors = [
            'input[type="checkbox"]',
            '[class*="agreement"] input',
            '[class*="protocol"] input',
            'input[name*="agreement"]',
            'input[name*="protocol"]'
        ]
        for selector in agreement_selectors:
            try:
                checkbox = page.locator(selector).first
                if checkbox.is_visible(timeout=1000) and not checkbox.is_checked():
                    checkbox.check()
                    print('已勾选协议复选框')
                    break
            except Exception:
                continue
    except Exception:
        pass

    # 验证表单是否已正确填写
    page.wait_for_timeout(500)
    try:
        account_value = account_input.input_value() if account_input else ''
        password_value = password_input.input_value() if password_input else ''
        brand_value = brand_input.input_value() if brand_input else ''
        print(f'表单验证 - 账号: {account_value[:3]}***, 密码: {"***" if password_value else "空"}, 品牌: {brand_value}')
    except Exception as e:
        print(f'表单验证失败: {e}')

    # 步骤3: 点击登录按钮
    login_button_selectors = [
        'button:has-text("登录")',
        'button:has-text("登陆")',
        'button[type="submit"]',
        'input[type="submit"]',
        'button.login',
        '.login-button',
        '[class*="login"] button',
        'button:has-text("Login")',
        'button:has-text("Sign in")',
        '[class*="login-btn"]',
        '[class*="submit"]'
    ]
    
    login_button = None
    for selector in login_button_selectors:
        try:
            login_button = page.locator(selector).first
            if login_button.is_visible(timeout=2000):
                print(f'找到登录按钮，准备点击')
                break
        except Exception:
            continue
    
    if login_button:
        # 等待按钮可点击
        login_button.wait_for(state='visible', timeout=5000)
        page.wait_for_timeout(500)
        
        # 点击登录按钮并等待导航（类似JavaScript版本的Promise.all）
        print('准备点击登录按钮并等待导航...')
        try:
            # 使用expect_navigation等待导航，使用更宽松的等待策略
            with page.expect_navigation(timeout=15000, wait_until='domcontentloaded'):
                login_button.click()
            print(f'登录按钮已点击，页面已跳转到: {page.url}')
        except Exception as e:
            print(f'等待导航超时: {e}，继续检查URL...')
            # 即使超时，也继续执行后续检查逻辑
            pass
    else:
        print('未找到登录按钮，尝试按Enter键')
        try:
            with page.expect_navigation(timeout=20000, wait_until='networkidle'):
                page.keyboard.press('Enter')
        except Exception:
            pass
    
    # 等待页面跳转，检查是否登录成功
    login_success = False
    max_wait_time = 30000  # 最多等待30秒
    import time
    start_time = time.time() * 1000
    
    # 先等待网络请求完成
    try:
        print('等待网络请求完成...')
        page.wait_for_load_state('networkidle', timeout=15000)
        print('网络请求已完成')
    except Exception as e:
        print(f'等待网络空闲超时: {e}')
    
    # 检查是否有错误提示或验证消息
    try:
        error_selectors = [
            '[class*="error"]',
            '[class*="alert"]',
            '[class*="message"]',
            '.ant-message',
            '.ant-notification',
            '[role="alert"]'
        ]
        for selector in error_selectors:
            try:
                error_element = page.locator(selector).first
                if error_element.is_visible(timeout=1000):
                    error_text = error_element.text_content()
                    if error_text:
                        print(f'检测到错误/提示信息: {error_text[:200]}')
            except Exception:
                continue
    except Exception:
        pass
    
    while (time.time() * 1000) - start_time < max_wait_time:
        current_url = page.url
        print(f'检查登录状态，当前URL: {current_url}')
        
        if '/page/login' not in current_url:
            login_success = True
            print(f'登录成功，页面已跳转到: {current_url}')
            break
        
        # 检查是否有错误提示
        try:
            page_text = page.text_content('body') or ''
            # 检查更具体的错误关键词
            error_keywords = ['错误', '失败', 'error', 'Error', '用户名或密码', '账号', '验证码', 'captcha']
            if page_text:
                for keyword in error_keywords:
                    if keyword.lower() in page_text.lower():
                        # 提取包含错误关键词的上下文
                        idx = page_text.lower().find(keyword.lower())
                        context = page_text[max(0, idx-50):idx+100]
                        print(f'检测到可能的相关信息: {context}')
        except Exception as e:
            print(f'检查页面内容时出错: {e}')
        
        # 等待1秒后再次检查
        page.wait_for_timeout(1000)
    
    if not login_success:
        final_url = page.url
        page_text = page.text_content('body') or ''
        print(f'登录超时，当前URL: {final_url}')
        print(f'页面内容预览: {page_text[:500] if page_text else "空内容"}')
        raise Exception('登录失败，页面仍在登录页面或登录超时')
    
    # 使用更宽松的等待策略
    try:
        page.wait_for_load_state('domcontentloaded', timeout=10000)
    except Exception as e:
        print(f'等待domcontentloaded超时，继续执行: {e}')
    
    page.wait_for_timeout(3000)
    print(f'登录后页面URL: {page.url}')


def verify_tenant_name(page: Page, expected_tenant_name: str = '合阔x'):
    """
    验证租户名
    
    Args:
        page: Playwright页面对象
        expected_tenant_name: 期望的租户名，默认为"合阔x"
    """
    page.wait_for_load_state('domcontentloaded')
    page.wait_for_timeout(2000)
    
    try:
        page_url = page.url
        print(f'当前页面URL: {page_url}')
        
        page.wait_for_load_state('domcontentloaded')
        page_title = page.title()
        print(f'当前页面标题: {page_title}')
    except Exception as e:
        print(f'获取页面信息失败: {e}')
    
    tenant_element = None
    found = False
    
    try:
        tenant_element = page.get_by_text(expected_tenant_name, exact=False).first
        tenant_element.wait_for(state='visible', timeout=5000)
        found = True
    except Exception:
        # 继续尝试其他方法
        pass
    
    if not found:
        try:
            header_selectors = ['header', 'nav', '[class*="header"]', '[class*="navbar"]']
            for header_selector in header_selectors:
                try:
                    header = page.locator(header_selector).first
                    if header.is_visible(timeout=2000):
                        tenant_element = header.get_by_text(expected_tenant_name, exact=False).first
                        if tenant_element.is_visible(timeout=2000):
                            found = True
                            break
                except Exception:
                    continue
        except Exception:
            print('在header中查找失败')
    
    if not found:
        page_text = page.text_content('body') or ''
        if expected_tenant_name in page_text:
            found = True
            print('租户名存在于页面文本中')
    
    # 使用标准断言验证租户名
    assert found is True, f'租户名 "{expected_tenant_name}" 未找到'
    
    # 验证页面文本中包含租户名（双重验证）
    page_text = page.text_content('body') or ''
    assert expected_tenant_name in page_text, f'页面文本中未包含租户名 "{expected_tenant_name}"'
    
    print(f'✓ 租户名验证通过: {expected_tenant_name}')

